include/ignition/gazebo/detail/BaseView.hh

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_GAZEBO_DETAIL_BASEVIEW_HH_
+#define IGNITION_GAZEBO_DETAIL_BASEVIEW_HH_
+
+#include <cstddef>
+#include <set>
+#include <unordered_map>
+#include <vector>
+
+#include "ignition/gazebo/Entity.hh"
+#include "ignition/gazebo/Types.hh"
+#include "ignition/gazebo/config.hh"
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace detail
+{
+/// \brief A key into the map of views
+using ComponentTypeKey = std::vector<ComponentTypeId>;
+
+/// \brief Hash functor for ComponentTypeKey
+struct ComponentTypeHasher
```

Comment:
> Can you add some comments here about the rationale for the chosen implementation? Maybe some external references?

Here are some references for the hash function I chose:
* https://stackoverflow.com/questions/20511347/a-good-hash-function-for-a-vector
* https://stackoverflow.com/questions/4948780/magic-number-in-boosthash-combine
* https://stackoverflow.com/questions/35985960/c-why-is-boosthash-combine-the-best-way-to-combine-hash-values
* https://softwareengineering.stackexchange.com/questions/402542/where-do-magic-hashing-constants-like-0x9e3779b9-and-0x9e3779b1-come-from

If anyone has thoughts for a better approach, I'd be open to hearing it. Otherwise, if this approach seems fine, I can add these links as documentation.

> It would also be interesting to write a simple unit test for this function that hashes a few component types and checks that they don't collide.

I can add this test coverage :+1: I'll do that once we settle on a hash function.

Reply:
> if this approach seems fine

I don't have a proposal for a better hash function

Reply:
Tests and references have been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa